reports/tests/test_api.py

```diff
@@ -113,8 +113,18 @@ def test_children_endpoint(user_api_client, snapshot, django_assert_max_num_quer
 
     ChildFactory()  # This is an orphan child so she should not be in the results
 
-    with django_assert_max_num_queries(6):
+    with django_assert_max_num_queries(10):
```

Comment:
Why was this needed to be increased? What change made this necessary?

Reply:
Good question. I wondered that my self too, but here are the queries made and I don't see anything wrong in them -- for some reason the `users_guardian` are selected too:

```
E               Failed: Expected to perform 6 queries or less but 10 were done
E               
E               Queries:
E               ========
E               
E               SELECT "users_user"."id", "users_user"."password", "users_user"."last_login", "users_user"."is_superuser", "users_user"."username", "users_user"."first_name", "users_user"."last_name", "users_user"."email", "users_user"."is_staff", "users_user"."is_active", "users_user"."date_joined", "users_user"."uuid", "users_user"."department_name" FROM "users_user" WHERE "users_user"."username" = 'test_user' LIMIT 21
E               
E               SELECT "django_content_type"."app_label", "auth_permission"."codename" FROM "auth_permission" INNER JOIN "users_user_user_permissions" ON ("auth_permission"."id" = "users_user_user_permissions"."permission_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "users_user_user_permissions"."user_id" = 3
E               
E               SELECT "django_content_type"."app_label", "auth_permission"."codename" FROM "auth_permission" INNER JOIN "auth_group_permissions" ON ("auth_permission"."id" = "auth_group_permissions"."permission_id") INNER JOIN "auth_group" ON ("auth_group_permissions"."group_id" = "auth_group"."id") INNER JOIN "users_user_groups" ON ("auth_group"."id" = "users_user_groups"."group_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "users_user_groups"."user_id" = 3
E               
E               SELECT "children_child"."created_at", "children_child"."updated_at", "children_child"."id", "children_child"."name", "children_child"."birthyear", "children_child"."postal_code", "children_child"."project_id" FROM "children_child" WHERE NOT (EXISTS(SELECT (1) AS "a" FROM "children_child" U0 LEFT OUTER JOIN "children_relationship" U1 ON (U0."id" = U1."child_id") WHERE (U1."guardian_id" IS NULL AND U0."id" = "children_child"."id") LIMIT 1)) ORDER BY "children_child"."name" ASC
E               
E               SELECT ("children_relationship"."child_id") AS "_prefetch_related_val_child_id", "users_guardian"."created_at", "users_guardian"."updated_at", "users_guardian"."id", "users_guardian"."user_id", "users_guardian"."first_name", "users_guardian"."last_name", "users_guardian"."phone_number", "users_guardian"."language", "users_guardian"."email" FROM "users_guardian" INNER JOIN "children_relationship" ON ("users_guardian"."id" = "children_relationship"."guardian_id") WHERE "children_relationship"."child_id" IN ('6a1143a9-8a08-4142-80b7-ec95dfe9c968'::uuid, '4cab1489-9a35-4d08-b497-7fd9ee74b7b6'::uuid, 'b71111a8-abb4-4e8a-b80c-d7a4262de2d8'::uuid, '4e1f2f18-ad56-48af-a68a-0bd32b4f17af'::uuid) ORDER BY "users_guardian"."created_at" DESC, "users_guardian"."last_name" ASC, "users_guardian"."first_name" ASC
E               
E               SELECT ("users_guardian_languages_spoken_at_home"."guardian_id") AS "_prefetch_related_val_guardian_id", "languages_language"."id", "languages_language"."alpha_3_code" FROM "languages_language" INNER JOIN "users_guardian_languages_spoken_at_home" ON ("languages_language"."id" = "users_guardian_languages_spoken_at_home"."language_id") WHERE "users_guardian_languages_spoken_at_home"."guardian_id" IN ('67433ef9-9416-487b-9046-6de3734d7de9'::uuid, '5520a0c5-993e-479c-af09-7150bc06e6ce'::uuid, '99d65d80-ecb1-4a36-9d7f-308b1beba92c'::uuid, '960c344c-bb46-4cbb-bfab-e5f2472ddbe9'::uuid, 'bab888ef-e813-4cb0-962e-8bfe74a54b62'::uuid) ORDER BY "languages_language"."alpha_3_code" ASC
E               
E               SELECT "users_guardian"."created_at", "users_guardian"."updated_at", "users_guardian"."id", "users_guardian"."user_id", "users_guardian"."first_name", "users_guardian"."last_name", "users_guardian"."phone_number", "users_guardian"."language", "users_guardian"."email" FROM "users_guardian" INNER JOIN "children_relationship" ON ("users_guardian"."id" = "children_relationship"."guardian_id") WHERE "children_relationship"."child_id" = '6a1143a9-8a08-4142-80b7-ec95dfe9c968'::uuid ORDER BY "users_guardian"."created_at" ASC LIMIT 1
E               
E               SELECT "users_guardian"."created_at", "users_guardian"."updated_at", "users_guardian"."id", "users_guardian"."user_id", "users_guardian"."first_name", "users_guardian"."last_name", "users_guardian"."phone_number", "users_guardian"."language", "users_guardian"."email" FROM "users_guardian" INNER JOIN "children_relationship" ON ("users_guardian"."id" = "children_relationship"."guardian_id") WHERE "children_relationship"."child_id" = '4cab1489-9a35-4d08-b497-7fd9ee74b7b6'::uuid ORDER BY "users_guardian"."created_at" ASC LIMIT 1
E               
E               SELECT "users_guardian"."created_at", "users_guardian"."updated_at", "users_guardian"."id", "users_guardian"."user_id", "users_guardian"."first_name", "users_guardian"."last_name", "users_guardian"."phone_number", "users_guardian"."language", "users_guardian"."email" FROM "users_guardian" INNER JOIN "children_relationship" ON ("users_guardian"."id" = "children_relationship"."guardian_id") WHERE "children_relationship"."child_id" = 'b71111a8-abb4-4e8a-b80c-d7a4262de2d8'::uuid ORDER BY "users_guardian"."created_at" ASC LIMIT 1
E               
E               SELECT "users_guardian"."created_at", "users_guardian"."updated_at", "users_guardian"."id", "users_guardian"."user_id", "users_guardian"."first_name", "users_guardian"."last_name", "users_guardian"."phone_number", "users_guardian"."language", "users_guardian"."email" FROM "users_guardian" INNER JOIN "children_relationship" ON ("users_guardian"."id" = "children_relationship"."guardian_id") WHERE "children_relationship"."child_id" = '4e1f2f18-ad56-48af-a68a-0bd32b4f17af'::uuid ORDER BY "users_guardian"."created_at" ASC LIMIT 1

../../../.pyenv/versions/3.9/envs/kukkuu-39/lib/python3.10/site-packages/pytest_django/fixtures.py:535: Failed
```

Before this PR the queries were these:

```
E               Queries:
E               ========
E               
E               SELECT "users_user"."id", "users_user"."password", "users_user"."last_login", "users_user"."is_superuser", "users_user"."username", "users_user"."first_name", "users_user"."last_name", "users_user"."email", "users_user"."is_staff", "users_user"."is_active", "users_user"."date_joined", "users_user"."uuid", "users_user"."department_name" FROM "users_user" WHERE "users_user"."username" = 'test_user' LIMIT 21
E               
E               SELECT "django_content_type"."app_label", "auth_permission"."codename" FROM "auth_permission" INNER JOIN "users_user_user_permissions" ON ("auth_permission"."id" = "users_user_user_permissions"."permission_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "users_user_user_permissions"."user_id" = 3
E               
E               SELECT "django_content_type"."app_label", "auth_permission"."codename" FROM "auth_permission" INNER JOIN "auth_group_permissions" ON ("auth_permission"."id" = "auth_group_permissions"."permission_id") INNER JOIN "auth_group" ON ("auth_group_permissions"."group_id" = "auth_group"."id") INNER JOIN "users_user_groups" ON ("auth_group"."id" = "users_user_groups"."group_id") INNER JOIN "django_content_type" ON ("auth_permission"."content_type_id" = "django_content_type"."id") WHERE "users_user_groups"."user_id" = 3
E               
E               SELECT "children_child"."created_at", "children_child"."updated_at", "children_child"."id", "children_child"."first_name", "children_child"."last_name", "children_child"."birthdate", "children_child"."postal_code", "children_child"."project_id" FROM "children_child" WHERE NOT (EXISTS(SELECT (1) AS "a" FROM "children_child" U0 LEFT OUTER JOIN "children_relationship" U1 ON (U0."id" = U1."child_id") WHERE (U1."guardian_id" IS NULL AND U0."id" = "children_child"."id") LIMIT 1)) ORDER BY "children_child"."last_name" ASC, "children_child"."first_name" ASC
E               
E               SELECT ("children_relationship"."child_id") AS "_prefetch_related_val_child_id", "users_guardian"."created_at", "users_guardian"."updated_at", "users_guardian"."id", "users_guardian"."user_id", "users_guardian"."first_name", "users_guardian"."last_name", "users_guardian"."phone_number", "users_guardian"."language", "users_guardian"."email" FROM "users_guardian" INNER JOIN "children_relationship" ON ("users_guardian"."id" = "children_relationship"."guardian_id") WHERE "children_relationship"."child_id" IN ('c4893e1d-f517-466c-ab03-5458f32328cd'::uuid, 'fc8d1cc6-d51f-42f9-b3b5-3fe430a9109a'::uuid, '1fc6c8bb-6504-4e57-9219-bf7bd42f0e95'::uuid, '5d57e776-87a6-4f7b-af60-b436c6fa4e28'::uuid)
E               
E               SELECT ("users_guardian_languages_spoken_at_home"."guardian_id") AS "_prefetch_related_val_guardian_id", "languages_language"."id", "languages_language"."alpha_3_code" FROM "languages_language" INNER JOIN "users_guardian_languages_spoken_at_home" ON ("languages_language"."id" = "users_guardian_languages_spoken_at_home"."language_id") WHERE "users_guardian_languages_spoken_at_home"."guardian_id" IN ('c6b14971-a6b5-4a29-9c53-c23da33e0293'::uuid, 'bf94921c-a3f6-4d58-807f-d3cb265835cd'::uuid, '2a4d1797-2412-4b37-93be-8e0a95c13472'::uuid, '515c06c5-efe2-4ff4-8699-b3db314b31e0'::uuid, '207bfa73-6957-4ec8-8dec-678570077a8f'::uuid) ORDER BY "languages_language"."alpha_3_code" ASC

```

Reply:
I think it must have something to do with the new ordering, but I don't understand why it would fetch them 1 by 1 with limit 1 and where it needs those.

Reply:
https://docs.djangoproject.com/en/5.0/ref/models/options/#ordering

<img width="892" alt="image" src="https://github.com/City-of-Helsinki/kukkuu/assets/389204/b60571b1-becb-4152-b4c5-27423a91717f">

Reply:
So it seems the queries were the same as before but now there is one query per child. But it could as well use the same kind of child_id in (...<array of child_id values>) like the one of the other queries does:

```sql
SELECT
	("children_relationship"."child_id") AS "_prefetch_related_val_child_id",
	"users_guardian"."created_at",
	"users_guardian"."updated_at",
	"users_guardian"."id",
	"users_guardian"."user_id",
	"users_guardian"."first_name",
	"users_guardian"."last_name",
	"users_guardian"."phone_number",
	"users_guardian"."language",
	"users_guardian"."email"
FROM "users_guardian"
INNER JOIN "children_relationship"
ON ("users_guardian"."id" = "children_relationship"."guardian_id")
WHERE "children_relationship"."child_id" IN (
	'6a1143a9-8a08-4142-80b7-ec95dfe9c968'::uuid,
	'4cab1489-9a35-4d08-b497-7fd9ee74b7b6'::uuid,
	'b71111a8-abb4-4e8a-b80c-d7a4262de2d8'::uuid,
	'4e1f2f18-ad56-48af-a68a-0bd32b4f17af'::uuid
)
ORDER BY
	"users_guardian"."created_at" DESC,
	"users_guardian"."last_name" ASC,
	"users_guardian"."first_name" ASC
```
Hmm... thinking about what could help here the Django query engine... any ideas? prefetch/select_related something?